tidy3d/plugins/microwave/path_integrals.py

```diff
@@ -0,0 +1,299 @@
+"""Helper classes for performing path integrals with fields on the Yee grid"""
+
+from __future__ import annotations
+
+import pydantic.v1 as pd
+import numpy as np
+from typing import Tuple
+from ...components.data.dataset import FieldDataset, FieldTimeDataset, ModeSolverDataset
+from ...components.data.data_array import AbstractSpatialDataArray
+from ...components.base import cached_property
+from ...components.types import Axis, Direction
+from ...components.geometry.base import Box
+from ...components.validators import assert_line, assert_plane
+from ...exceptions import DataError
+
+
+class AbstractAxesRH:
+    """Represents an axis-aligned right-handed coordinate system with one axis preferred."""
+
+    @cached_property
+    def main_axis(self) -> Axis:
+        """Subclasses should implement this method."""
+        raise NotImplementedError()
+
+    @cached_property
+    def remaining_axes(self) -> Tuple[Axis, Axis]:
+        """Axes in plane, ordered to maintain a right-handed coordinate system"""
+        axes = [0, 1, 2]
+        axes.pop(self.main_axis)
+        if self.main_axis == 1:
+            return (axes[1], axes[0])
+        else:
+            return (axes[0], axes[1])
+
+
+class AxisAlignedPathIntegral(AbstractAxesRH, Box):
+    """Class for defining the simplest type of path integral which is aligned with Cartesian axes."""
+
+    _line_validator = assert_line()
+
+    extrapolate_to_endpoints: bool = pd.Field(
+        False,
+        title="Extrapolate to endpoints",
+        description="If the endpoints of the path integral terminate at or near a material interface, "
+        "the field is likely discontinuous. This option ignores fields outside and on the bounds "
+        "of the integral. Should be turned on when computing voltage between two conductors. ",
+    )
+
+    snap_path_to_grid: bool = pd.Field(
+        False,
+        title="Snap path to grid",
+        description="It might be desireable to integrate exactly along the Yee grid associated with "
+        "a field. If enabled, the integration path will be snapped to the grid.",
+    )
+
+    def compute_integral(self, scalar_field: AbstractSpatialDataArray):
+        """Computes the defined integral given the input `scalar_field`."""
```

Comment:
for docstrings, we need to use double backslash quotes for variables. Because it gets interpreted as restructured text (RST) in the docs. Be sure to fix if it comes up in other spots too